aten/src/ATen/native/AdaptiveMaxPooling3d.cpp

```diff
@@ -0,0 +1,402 @@
+          const Tensor& input,
+          IntArrayRef output_size)
+{
+  int dimD = 0;
+  int dimT = 1;
+  int dimH = 2;
+  int dimW = 3;
+  int64_t sizeB = 1;
+  int64_t sizeD = 0;
+  int64_t isizeT = 0;
+  int64_t isizeH = 0;
+  int64_t isizeW = 0;
+
+  int64_t istrideB = 0;
+  int64_t istrideD = 0;
+  int64_t istrideT = 0;
+  int64_t istrideH = 0;
+  int64_t istrideW = 0;
+
+  for (int64_t i = 0; i < input.ndimension(); i++) {
+    AT_CHECK(input.size(i) > 0,
+      "adaptive_max_pool3d: expected input to have non-empty spatial dimensions, "
+      "but input has sizes ", input.sizes(), " with dimension ", i, " being "
+      "empty");
+  }
+
+  AT_CHECK((input.ndimension() == 4 || input.ndimension() == 5),
+    "non-empty 4D or 5D (batch mode) tensor expected for input");
+
+  // the jit sometimes passes output_size.size() == 1
+  AT_CHECK(output_size.size() == 1 || output_size.size() == 3,
+    "adaptive_max_pool3d: internal error: output_size.size() must be 1 or 3");
+
+  if (input.ndimension() == 5)
+  {
+    istrideB = input.stride(0);
+    sizeB = input.size(0);
+    dimD++;
+    dimT++;
+    dimH++;
+    dimW++;
+  }
+
+  /* sizes */
+  sizeD  = input.size(dimD);
+  isizeT = input.size(dimT);
+  isizeH = input.size(dimH);
+  isizeW = input.size(dimW);
+  /* strides */
+  istrideD = input.stride(dimD);
+  istrideT = input.stride(dimT);
+  istrideH = input.stride(dimH);
+  istrideW = input.stride(dimW);
+
+  int64_t osizeT = output_size[0];
+  int64_t osizeH = output_size.size() == 1 ? output_size[0] : output_size[1];
+  int64_t osizeW = output_size.size() == 1 ? output_size[0] : output_size[2];
+
+  /* resize output */
+  if (input.ndimension() == 4)
+  {
+    output.resize_({sizeD, osizeT, osizeH, osizeW});
+    /* indices will contain max input locations for each output point */
+    indices.resize_({sizeD, osizeT, osizeH, osizeW});
+
+    AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool3d_cpu", [&] {
+      auto input_data = input.data<scalar_t>();
+      auto output_data = output.data<scalar_t>();
+      auto indices_data = indices.data<int64_t>();
+
+      adaptive_max_pool3d_out_frame<scalar_t>(input_data, output_data,
+                                              indices_data,
+                                              sizeD,
+                                              isizeT, isizeH, isizeW,
+                                              osizeT, osizeH, osizeW,
+                                              istrideD, istrideT,
+                                              istrideH, istrideW);
+      }
+    );
+  }
+  else
+  {
+    int64_t b;
+
+    output.resize_({sizeB, sizeD, osizeT, osizeH, osizeW});
+    /* indices will contain max input locations for each output point */
+    indices.resize_({sizeB, sizeD, osizeT, osizeH, osizeW});
+
+#pragma omp parallel for private(b)
+    for (b = 0; b < sizeB; b++)
+    {
+      AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool3d_cpu", [&] {
```

Comment:
Oops, you're vdispatching inside the OMP loop! Please check https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide#dispatch-and-openmp for guidance!

Reply:
cc @xmnlab

Reply:
Hmm, I think I cargo-culted that one from `AdaptiveAveragePooling.cpp`.  I'll read the Wiki more closely again.